Collections changes feature branch

daltojohnso opened the pull request:
This is the feature branch for the work in #19758.

The primary pieces of this feature branch have been reviewed in separate PRs, with a few exceptions for a handful of small PRs at the end that I will specify.

- Add `PinnedItemCard`: https://github.com/metabase/metabase/pull/19765
- Add `PinnedItemOverview` and replace the existing pinned items table: https://github.com/metabase/metabase/pull/19787
- Miscellaneous UI changes to collections page: https://github.com/metabase/metabase/pull/19792
- Show card visualizations in pinned overview: https://github.com/metabase/metabase/pull/19813
- Make pinned cards draggable and droppable into the collections sidebar: https://github.com/metabase/metabase/pull/19829
- (**unreviewed**) unit & e2e tests for the new pinned overview section and ui changes: https://github.com/metabase/metabase/pull/19851
- (**unreviewed**) a changed empty state: https://github.com/metabase/metabase/pull/19857
- (**unreviewed**) some additional styling tweaks from @mazameli: https://github.com/metabase/metabase/pull/19853

**Testing**
- create a new collection. The empty state should show a new banner informing the user that they can pin items. 
- add questions and dashboards and models to the collection. The banner should persist until the user has pinned an item.
- pin a dashboard. A card for the dashboard should show up. The card should be draggable to the left sidebar to move it to a different folder. Clicking the card should take the user to the dashboard. Hovering over the card should show an action menu in the top right corner of the card. Test all the various actions, unpinning, duplicating, archiving, and moving. 
- pin a model. It should look very similar to the dashboard card except that it is separated by some label text. Should work the same. Clicking should take user to the model in the query builder. 
- test long descriptions and names for dashboards and models
- Pin a question. The visualization for the question should show sorta like if we were on a dashboard. The entire viz should be clickable to take the user to the saved question. An action menu should show in the top right corner like the dashboard and model cards. Dragging the visualization should show a card for performance reasons 
- the above items should be ordered alphabetically

daltojohnso committed:
Add PinnedItemCard component (#19765)

* convert Card component to typescript

* add open state class to EntityMenu

* Add PinnedItemCard component

* prevent the bubbling of EntityMenu events

* replace class with styled-component stuff

* define a const

daltojohnso committed:
Replace pinned collection items table with pinned overview section (#19787)

* Add PinnedItemOverview component

* Replace the pinned table with overview

* add types to collection utils

* fix pinning + types

* remvoe the old PinnedItemsTable

* fix/remove e2e tests that no longer make sense

* replace 'Datasets' with 'Models'

* Update frontend/src/metabase/collections/components/PinnedItemOverview/PinnedItemOverview.tsx

Co-authored-by: Maz Ameli <maz@metabase.com>

* add unique text for root collection

Co-authored-by: Maz Ameli <maz@metabase.com>

daltojohnso committed:
Collections page UI changes (#19792)

* make a bunch of ui tweaks to collections

* resize Titles for PinnedItemCards

* resize collection description headings

* Add tooltip to pin button

* Fix unit test

* fix e2e tests

* add a default description

* Tbody --> TBody

* Hide long titles & add tooltip to show entire title

Co-authored-by: Maz Ameli <maz@metabase.com>

daltojohnso committed:
Add empty pinned items section banner (#19857)

Fix e2e test

don't show empty states when loading

styling tweaks

Fix test

daltojohnso committed:
Add card visualizations to the pinned section of collections (#19813)

* First pass at card viz in collections

* minor style fixes

* move redux stuff up to CollectionContent

* padding tweaks

* Make the entire card clickable

* add generic error message

* Add propTypes

* a little cleanup

* create ActionMenu component

* Fix pin action fn

* Fix circular dep & storybook story

* actually import the function

* add comment // use EventSandbox

* Add preventDefault prop to avoid breaking things

daltojohnso committed:
Make pinned cards drag-and-droppable into the collections sidebar (#19829)

* first pass at dnd for pinned cards

* use cards instead of viz for performance

* fix type

daltojohnso committed:
Add tests for collection changes (#19851)

* first pass at dnd for pinned cards

* use cards instead of viz for performance

* fix type

* Add tests for collection changes

daltojohnso committed:
fix Button import

kulyk commented:
This looks fantastic @daltojohnso 😍 I have a few questions about the changes:

**1. Empty collection state banner**

> create a new collection. The empty state should show a new banner informing the user that they can pin items.

Should I see this one (from Figma)?

<img width="512" alt="CleanShot 2022-01-25 at 13 42 09@2x" src="https://user-images.githubusercontent.com/17258145/150970773-4bd07afb-8fae-425f-99e1-1a1fa78f2eb8.png">

Because currently, I can only see the usual empty state

<details>
<summary>Screenshot</summary>

<img width="980" alt="CleanShot 2022-01-25 at 13 42 43@2x" src="https://user-images.githubusercontent.com/17258145/150970941-f88e4f4b-8f78-4a7e-8926-81864a5bc237.png">

</details>

**2. Drag-n-drop for models** 

DND works great, though it seems like it doesn't work for models. See the video:

<details>
<summary>Demo video</summary>

https://user-images.githubusercontent.com/17258145/150971191-7b0c117c-ee9d-4b93-80a6-97e031d5c9c4.mp4

</details>

**3. Unpinned items' icon colors** 

In Figma model and dashboard icons are blue and question icons are grey, but on the branch, everything is blue

kulyk commented:
Re: collection empty state banner. #19857 uses "dataset" instead of "models" (see [a comment on the PR](https://github.com/metabase/metabase/pull/19857#discussion_r791638267)). Could you please fix that?

daltojohnso committed:
remove async

daltojohnso committed:
datasets --> models

daltojohnso committed:
fix dragging data models

daltojohnso commented on frontend/src/metabase/collections/components/PinnedItemCard/PinnedItemCard.unit.spec.js:
```diff
@@ -0,0 +1,84 @@
+    const { container } = setup();
+    userEvent.click(container.querySelector(MENU_ICON_SELECTOR));
+    expect(screen.getByText("Unpin this item")).toBeInTheDocument();
+  });
```
Removed

daltojohnso committed:
Make question icons in collections bg-dark

daltojohnso commented:
Thanks for the thorough review @kulyk and catching that clobbered commit! Oof. I've added it back, fixed model dragging, and fixed the copy regarding datasets, and fixed the color of question icons.

kulyk reviewed (approved):
Thanks @daltojohnso 🏆 Let's ship it!
